ip_country/src/country_block_serde.rs

```diff
@@ -341,119 +528,94 @@ impl CountryBlockDeserializer for CountryBlockDeserializerIpv6 {
-            return None;
-        }
-        differences
-            .into_iter()
-            .for_each(|d| segments[d.index] = d.value as u16);
-        Some(StreamRecordIpv6 {
-            start: Ipv6Addr::from(segments),
-            country_idx: bit_queue.take_bits(9)? as usize,
-        })
+            .collect()
     }
 }
 
-#[derive(Debug)]
-struct StreamRecordIpv4 {
-    start: Ipv4Addr,
-    country_idx: usize,
-}
-
-struct StreamRecordIpv6 {
-    start: Ipv6Addr,
-    country_idx: usize,
-}
-
-struct Difference {
-    index: usize,
-    value: u64,
+struct VersionedIP<IPType, SegmentNumRep, const SEGMENTS_COUNT: usize> {
+    ip: IPType,
+    segment_num_rep: PhantomData<SegmentNumRep>,
 }
 
-fn plus_one_ipv4(ip_addr: Ipv4Addr) -> Ipv4Addr {
-    let old_data = u32_from_ipv4(ip_addr);
-    let new_data = old_data.overflowing_add(1).0;
-    Ipv4Addr::from(new_data)
+impl<IPType, SegmentNumRep, const SEGMENTS_COUNT: usize>
+    VersionedIP<IPType, SegmentNumRep, SEGMENTS_COUNT>
+{
+    fn new(ip: IPType) -> VersionedIP<IPType, SegmentNumRep, SEGMENTS_COUNT> {
+        let segment_num_rep = Default::default();
+        Self {
+            ip,
+            segment_num_rep,
+        }
+    }
 }
 
-fn minus_one_ipv4(ip_addr: Ipv4Addr) -> Ipv4Addr {
-    let old_data = u32_from_ipv4(ip_addr);
-    let new_data = old_data.overflowing_sub(1).0;
-    Ipv4Addr::from(new_data)
+impl<IPType: Debug, SegmentNumRep, const SEGMENTS_COUNT: usize> Debug
+    for VersionedIP<IPType, SegmentNumRep, SEGMENTS_COUNT>
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.write_fmt(format_args!("VersionedIP {{ ip: {:?} }}", self.ip))
+    }
 }
 
-fn u32_from_ipv4(ip_addr: Ipv4Addr) -> u32 {
-    let octets = ip_addr.octets();
-    let mut bit_data = 0u32;
-    octets.into_iter().for_each(|octet| {
-        bit_data <<= 8;
-        bit_data |= octet as u32;
-    });
-    bit_data
+#[derive(Debug)]
+struct StreamRecord<IPType: Display, SegmentNumRep, const SEGMENTS_COUNT: usize> {
+    start: VersionedIP<IPType, SegmentNumRep, SEGMENTS_COUNT>,
+    country_idx: usize,
 }
 
-fn plus_one_ipv6(ip_addr: Ipv6Addr) -> Ipv6Addr {
-    let old_data = u128_from_ipv6(ip_addr);
-    let new_data = old_data.overflowing_add(1).0;
-    Ipv6Addr::from(new_data)
+impl<IPType, SegmentNumRep, const SEGMENTS_COUNT: usize>
+    StreamRecord<IPType, SegmentNumRep, SEGMENTS_COUNT>
+where
+    IPType: From<[SegmentNumRep; SEGMENTS_COUNT]> + Display,
+    SegmentNumRep: TryFrom<u64>,
+    <SegmentNumRep>::Error: Debug,
+{
+    fn new(
+        differences: Vec<Difference>,
+        mut segments: [SegmentNumRep; SEGMENTS_COUNT],
+        country_idx: usize,
+    ) -> StreamRecord<IPType, SegmentNumRep, SEGMENTS_COUNT> {
+        differences.into_iter().for_each(|d| {
+            segments[d.index] = SegmentNumRep::try_from(d.value)
+                .expect("Difference noted in a bigger number than which the IP segment can contain")
```

Comment:
Please modify the message as follows 

Difference noted in a bigger number than which  ---> Difference represented by a bigger number than which